flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/schema/ExpandingPreparingTable.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.schema;
+
+import org.apache.flink.table.planner.plan.stats.FlinkStatistic;
+
+import org.apache.calcite.plan.RelOptSchema;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelShuttleImpl;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.type.RelDataType;
+
+import javax.annotation.Nullable;
+
+/**
+ * A common subclass for all tables that are expanded to sub-query (views). It handles
+ * recursive expanding of sub-views automatically.
+ */
+public abstract class ExpandingPreparingTable extends FlinkPreparingTableBase {
+	protected ExpandingPreparingTable(
+		@Nullable RelOptSchema relOptSchema,
+		RelDataType rowType,
+		Iterable<String> names,
+		FlinkStatistic statistic) {
+		super(relOptSchema, rowType, names, statistic);
+	}
+
+	/**
+	 * Expands at least the most outer level of a relational tree. Does not need to expand
+	 * any nested scans of an {@link ExpandingPreparingTable}. Those will be expanded recursively.
```

Comment:
at least -> at most ?

Reply:
This is correct. The implementation is free to expand more than the outer level. Than the `ExpandingPreparingTable` will make an empty pass over the nested nodes, cause inner levels were already expanded. This is absolutely fine though.

Reply:
Yes, the logic is right, but a little conflict with the method name, which is a little confusing.

Reply:
Is it though? The method says expand the top level. That's what the implementer should do. That's the contract of the method. The comment just says its fine if you do more.

Reply:
I'm fine changing the name to sth different if you have a good suggestion. The change `at least` to `at most` is not valid though.

Reply:
Or, better, the sub-class just return an un-expand relational node, let's do the expanding work all in the parent class, which would make the interface duty/function more clear.

Reply:
Yes, that's what the method does. By expanding I meant:
- in case of sql string view it means converting sql to relNode
- in case of Operation view it means converting SqlOperation to relNode (this will actually expand all nested scans because we are using expanding scan factory - that's why I added the comment)

If it makes it more understandable how about renaming it to `convertToRel`? Any other ideas for a name?

Reply:
Yes, that's what the method does. By expanding I meant:
- in case of sql string view it means converting sql to relNode
- in case of Operation view it means converting SqlOperation to relNode (this will actually expand all nested scans because we are using expanding scan factory - that's why I added the comment)

If it makes it more understandable how about renaming it to `convertToRel`? Any other ideas for a name?

Reply:
I'm okey with the new name.